### DISCUSSION TITLE
Migrating from yarn-offline-mirror

### DISCUSSION THREAD

**Discussion by Hotmonkey** [pnpm/pnpm#15938] [Ideas]
I have a project using yarn offline mode for CI build. I'm trying to migrate this project from yarn to pnpm same in offline mode. But I was blocked by the following issues:

1. **How to generate pnpm-lock.yaml from yarn.lock?** I tried to convert yarn.lock to package-lock.json first, and then run _pnpm import_ to generate the pnpm-lock,yaml. Unfortunately, some tgz tarballs in repo doesn't exist in our own registry any more, _pnpm import_ will fire the error that the deleted package versions are unavailable. But the deleted packages are in offline-mirror directory. Is that possible to force pnpm searching the packages from offline-mirror directory first?
2. **How to install from offline-mirror first?** It seems that _pnpm install --offline_ doesn't support configure the offline-mirror directory with tgz tarballs. Is that possible to configure that?
3. **How to auto add offline tarball into offline-mirror directory?** Is there a way for _pnpm add_ to work like _yarn add_ after configured offline mode: the new tarballs will be auto put into offline-mirror directory? So that I don't need to put the tarball into offline-mirror every time adding new packages.

I went through pnpm docs, didn't find the solutions to my issues. Is that possible to make that?

**Comment by zkochan** [comment]
> How to generate pnpm-lock.yaml from yarn.lock?

That is currently not possible but here's a PR that tries to add that feature: https://github.com/pnpm/pnpm/pull/2836

> How to install from offline-mirror first? 

pnpm doesn't have the offline-mirror feature. It doesn't even write the tarballs to the filesystem. But you can commit the store to the repository, if you want to install with `--offline` on CI. So set the `store-dir` location to something like `pnpm-store` in the `.npmrc` of the project and run `pnpm install`.

**Comment by xiaoxiangmoe** [comment]
However, the yarn offline mirror with tgz tarballs is more readable than the pnpm store-dir when git check in.

It is also easier for us to write code to verify the resolution.integrity of the tarballs than pnpm store-dir.

**Comment by zkochan** [comment]
Closing: current pnpm covers this.

`pnpm import` reads yarn.lock, package-lock.json and npm-shrinkwrap.json directly. For the offline half, a warm store plus `pnpm install --offline` replaces the yarn offline mirror.

If you still hit this on the latest pnpm, please comment with a reproduction and we will reopen.

---
Written by an agent (Claude Code, claude-opus-5).

**Comment by andersk** [comment]
@zkochan What’s still missing is the ability to _create_ a warm store from tarballs that the user downloaded separately from the registry.

Why this is important: hermetic build systems like Flatpak and Nix require that all downloads are declared in advance with hashes (to improve supply chain security, source provenance, reproducibility, and caching), and strictly enforce this with build sandboxing.

The way pnpm support was added to Flatpak (flatpak/flatpak-builder-tools#511, flatpak/flatpak-builder-tools#537) required many hundreds of lines of Python that replicates construction of the SQLite database and manually reimplements a subset of the msgpackr binary format, with different paths for each pnpm store version. This is a huge maintenance headache resulting in long delays before new pnpm versions can be used in Flatpak.

Meanwhile Nix, with the flexibility of fixed-output derivations that Flatpak lacks, uses `pnpm` to fetch everything from the registry and hashes a tarball of the entire pnpm store as a single unit ([code](https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/node/fetch-pnpm-deps/default.nix)). But this still requires direct SQLite manipulation since the SQLite binary format is not deterministic. And it’s wasteful since it thwarts the consolidation of multiple downloads and duplicate cache usage for common dependencies shared by multiple packages.

Both systems would be simplified and improved by the ability to initialize a warm store from an unstructured collection of the required tarballs.

**Comment by zkochan** [comment]
right, ok, we can support that probably

**Comment by altaywtf** [comment]
I'm working on a fix for this.

Written by an agent (Claude Code, Opus 5.5).

**Comment by altaywtf** [comment]
A fix is ready for review in https://github.com/pnpm/pnpm/pull/15844. CI is green.

Written by an agent (Claude Code, Opus 5.5).
